osd: Remove leaked clone objects (SnapMapper malformed key)

Matan-B opened the pull request:
This PR aims to to solve a clone objects leak to clusters affected by: https://tracker.ceph.com/issues/56147

* Step 1: (for each affected osd)
Using a new asock command (`fix_malformed_snapmapper_keys`) that converts the malformed SnapMapper keys to the correct structure. 
To cover clone objects that belong to an EC pool, all possible shard prefixes are inserted. (Total of 128 keys).

* Step 2: (mon command) 
**Separated into #53235**
Since the correct key exists, `ceph osd pool force_reremove_snap <pool> <lower bound> <upper bound>` mon command can be used to remove any non-existing (already removed) snapshots. 

* Step 3: (for each affected osd)
Clean up the extra inserted keys in step 1 after re-deleting the snapshots by using (`cleanup_snapmapper_possible_keys`)

---

The leaked clone objects associated to the re-deleted snapshot will be removed once the key is fixed.

```
2023-08-08T17:43:24.928+0000 7fcb7e0f4700 10 snap_mapper.convert_malformed
2023-08-08T17:43:24.928+0000 7fcb7e0f4700 20 snap_mapper.convert_malformed old key: SNA_2_0000000000000001_
2023-08-08T17:43:24.928+0000 7fcb7e0f4700 20 snap_mapper.convert_malformed converted key: SNA_2_0000000000000001_0000000000000002.ECF34CE6.1.objectone..                                                                           
2023-08-08T17:43:24.929+0000 7fcb7e0f4700 10 snap_mapper.convert_malformed converted 1 keys
2023-08-08T17:43:24.929+0000 7fcb7e0f4700  1 snap_mapper.convert_malformed converted 1 keys in 0.000145347s
```

TO DO:
- [x] self-managed snaps
- [x] EC pool support
- [x] Remove additional keys inserted after snapshot is removed
- [x] Release osd_lock on SnapMapper traverse
- [ ] tests


Fixes: https://tracker.ceph.com/issues/62596

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

ronen-fr commented on src/librados/IoCtxImpl.cc:
```diff
@@ -422,6 +422,20 @@ int librados::IoCtxImpl::snap_remove(const char *snapName)
 
+int librados::IoCtxImpl::snap_remove_again()
+{
+  int reply;
```
move the definition downwards

ronen-fr commented on src/mon/OSDMonitor.cc:
```diff
@@ -14264,6 +14286,27 @@ bool OSDMonitor::prepare_pool_op(MonOpRequestRef op)
+        //
+        // Note: for this to work with selfmanaged snaps IoCtx::snap_list should include
+        //       self-managed snaps. Currently the iteration is up to last snapid (pool snaps)
+        for (uint64_t i = 1; i < pp.get_snap_seq() ; i++) {
```
is u64 the correct choice here? how far can snap_seq reach?
I'd use a signed integer, so that - if get_snap_seq) would ever be modified to return an error
as a negative number, we won't crash here

Matan-B replied:
get_snap_seq() is a const getter for:
```
snapid_t snap_seq = 0;        ///< seq for per-pool snapshot
```
 > is u64 the correct choice here? how far can snap_seq reach?

```
struct snapid_t {
  uint64_t val;
   ..
}
```

ronen-fr commented on src/osd/SnapMapper.cc:
```diff
@@ -220,6 +220,12 @@ string SnapMapper::to_raw_key(
   return get_prefix(in.second.pool, in.first) + shard_prefix + in.second.to_str();
 }
 
+string SnapMapper::to_raw_key_no_shard_prefix(
```
do use std::, as the other functions do

ronen-fr commented on src/osd/SnapMapper.cc:
```diff
@@ -881,6 +887,11 @@ bool SnapMapper::is_legacy_mapping(const string &to_test)
     LEGACY_MAPPING_PREFIX;
 }
 
+bool SnapMapper::is_malformed_mapping(const string &to_test)
```
(string_view to_test)

Matan-B replied:
This function calls is_mapping() which gets const string&, I don't want to change any existing code to avoid conflicts if this will be backported.

ronen-fr commented on src/osd/SnapMapper.cc:
```diff
@@ -972,4 +992,65 @@ int SnapMapper::convert_legacy(
+  ObjectStore *store,
+  ObjectStore::CollectionHandle& ch,
+  ghobject_t hoid,
+  unsigned max)
```
please don't use 'max':
a) it's not descriptive enough, and
b) greps with the max() function/macro

ronen-fr commented on src/osd/SnapMapper.cc:
```diff
@@ -972,4 +992,65 @@ int SnapMapper::convert_legacy(
+  auto start = ceph::mono_clock::now();
+
+  iter->upper_bound(SnapMapper::MAPPING_PREFIX);
+  map<string,ceph::buffer::list> to_set;
```
'g, c'

Matan-B replied:
What is 'g, c'?

ronen-fr commented on src/osd/SnapMapper.cc:
```diff
@@ -972,4 +992,65 @@ int SnapMapper::convert_legacy(
+  iter->upper_bound(SnapMapper::MAPPING_PREFIX);
+  map<string,ceph::buffer::list> to_set;
+  while (iter->valid()) {
+    bool valid = SnapMapper::is_malformed_mapping(iter->key());
```
A confusing collision of naming:
'is_malformed..()' returns true for 'valid' (not malformed)?

ronen-fr commented on src/osd/SnapMapper.cc:
```diff
@@ -972,4 +992,65 @@ int SnapMapper::convert_legacy(
+  while (iter->valid()) {
+    bool valid = SnapMapper::is_malformed_mapping(iter->key());
+    if (valid) {
+      dout(20) << __func__ << "old key: " << iter->key() << dendl;
```
can be combined with the next log (and maybe use fmtlib?)

ronen-fr commented on src/osd/osd_types.cc:
```diff
@@ -1723,6 +1723,17 @@ snapid_t pg_pool_t::snap_exists(std::string_view s) const
+
+bool pg_pool_t::snap_exists(snapid_t s) const
+{
+  for (auto p = snaps.cbegin(); p != snaps.cend(); ++p) {
```
auto& ?

ronen-fr commented on src/osdc/Objecter.cc:
```diff
@@ -3973,6 +3973,29 @@ void Objecter::delete_pool_snap(
+  decltype(PoolOp::onfinish)&& onfinish)
+{
+  unique_lock wl(rwlock);
+  ldout(cct, 10) << "delete_pool_snap; pool: " << pool << dendl;
```
Can the log be moved to before the lock?

Matan-B replied:
removed

ronen-fr commented:
> Do you find any design flaws with this approach?

Working on that...

athanatos commented:
I don't want to add a librados api for this purpose.  librados api commands need to be maintained long term and shouldn't be created for one-off fixes.  Instead, I'd suggest a mon command.  See OSDMonitor::prepare_command_impl, search for "pool rmsnap".  We probably want an optional argument for "pool rmsnap" that forces it be placed back into the OSDMap.

Matan-B commented on src/mon/OSDMonitor.cc:
```diff
@@ -14264,6 +14286,27 @@ bool OSDMonitor::prepare_pool_op(MonOpRequestRef op)
+          // exclude existing snapshots
+          if (!pp.snap_exists(i)) {
+            dout(10) << "removing snap " << i << " again. pool " << m->pool << dendl;
+	        pending_inc.new_removed_snaps[m->pool].insert(i);
```
removed

Matan-B commented:
> I don't want to add a librados api for this purpose. librados api commands need to be maintained long term and shouldn't be created for one-off fixes. Instead, I'd suggest a mon command.
 ~See OSDMonitor::prepare_command_impl, search for "pool rmsnap". We probably want an optional argument for "pool rmsnap" that forces it be placed back into the OSDMap.~ (edit: "pool rmsnap" is specific to pool snaps, so we'll need a new additional command) You should be able to add a a command to OSDMonitor::prepare_command/preprocess_command like "pool rmsnap" but for an unmanaged snap.

Moved to a new mon command.

>See prepare_pool_op/preprocess_pool_op handlers for POOL_OP_DELETE_SNAP for the monitor side of the librados self-managed snapshot removal logic.

Pool snaps are ok since we have the existing snaps at hand. However, for self-managed ones I couldn't find an elegant way to obtain the existing/purged snaps.
`PGMapDigest::purged_snaps` is possible to get hold of although it only holds the purged snap ids without the corresponding pool.
`pg_info_t` also holds complete purged snaps list but I couldn't obtain it from the monitor side. Moreover, traversing through all of the pgs to collect the purged snaps might be too expensive.

I think that the best option to get the existing snap ids is to let the user specify them as a parameter.
This part will handled by a (future) script that will also remove the snapshots in portions.

> As an aside, we do probably want there to be a script which re-removes snaps in smallish batches (probably specified as an argument to the above command) which are confirmed to be trimmed before moving onto the next one. As such, the command should either take a snapshot to remove or a set (comma delimited?). That avoids needing to re-trim all of the previously removed snapshots on a cluster at once, and allows specifically re-removing as little as a single snap.

Added a lower and upper snap id bounds.
Note: We avoid removing not yet taken snapshots (UB) even if the upper snap id bound passed exceeds it.

---

Example usage for rbd:
```
$ rbd snap ls <pool>/<image>                                                                                                                                                                                                                                                   
SNAPID  NAME   SIZE   PROTECTED  TIMESTAMP                                                                                                                                                                                             
     5  snp14  4 KiB             Mon Aug 21 11:47:07 2023                                                                                                                                                                              
     6  snp15  4 KiB             Mon Aug 21 11:47:09 2023                                                                                                                                                                              

$ ceph osd pool rmsnap_again <pool> 0 10 5,6                                                                                                                              
                                                                                                                   
removing snap 1 again from pool 2
removing snap 2 again from pool 2
removing snap 3 again from pool 2
removing snap 4 again from pool 2
snap 5 was specified and won't be removed
snap 6 was specified and won't be removed
```

`ceph osd pool rmsnap_again <pool> <lower_bound> <higher_bound> <existing snap ids>`

Matan-B commented:
@athanatos, I updated the monitor command to simply iterate up to the latest `snap_seq` (maintained in both snapshots types) and to _redelete_ only the purged snapshots which are obtained from the monitor db.

Note: In earlier discussions I assumed that the purged snaps kv entries are not complete. However, that seem to be wrong as they are not being trimmed in any way.

athanatos commented on src/osd/OSD.cc:
```diff
@@ -3093,6 +3093,22 @@ will start to track new ops received afterwards.";
     pg_recovery_stats.reset();
   }
 
```
I don't understand how this command is supposed to work.  Add brief, precise comments to the header declarations of convert_malformed_key and convert_malformed in the format:

```
/**
 * method_name
 *
 * Explanation of what method_name *does*, not how it works.  A reader should be able
 * to use this method correctly by reading this comment without needing to read the
 * implementation.
 */
```

If you need to elaborate on something in the implementation, do it in the definition body.

From what I can tell from the implementation, it looks like the goal here is to remove the malformed key and write back (for an ec pool) all 128 possible legacy keys.  What's the plan for ultimately removing the extras?  You could avoid this by passing in a set of all spg_t and pg_pool_t (for pg_num so you can run stable_mod) on the OSD.  Virtually every time, a particular PG will have only one shard present on the OSD, so you can take the hobject_t, map it to a pg, find the matching spg_t, and use that to determine the shard.  You could simply skip any objects in PGs that happen to have multiple shards present.

athanatos replied:
Also, let's name it "fix_malformed_snapmapper_keys".  The description should reference the bug id and be more useful: "Scans the SnapMapper for malformed keys created by bug <bugid> during upgrade from <release> to <other release>."

athanatos replied:
I'm worried about holding the osd lock during the scan as well.  While you have bounded the number of keys to remove, you're going to keep re-reading the SnapMapper up to that point.  Is it the case that the malformed keys always sort at the start of the region?  If so, that's an invariant you should explain in the implementation definition (not the header).

athanatos replied:
The commit message needs to reference the bug and give a very brief description of the history and how this command is supposed to remedy it.

athanatos replied:
You also need to check for and avoid overwriting the fixed key if it already exists.  During the original upgrade that caused the problem and now, it's possible that some of the objects had their SnapMapper keys rewritten and thereby fixed.  Overwriting them with the old value would be incorrect.

athanatos commented on src/mon/MonCommands.h:
```diff
@@ -1071,6 +1071,12 @@ COMMAND("osd pool rmsnap "
 	"name=pool,type=CephPoolname "
 	"name=snap,type=CephString",
 	"remove snapshot <snap> from <pool>", "osd", "rw")
+COMMAND("osd pool rmsnap_again "
```
How about "osd pool force_reremove_snap"?

athanatos commented on src/mon/MonCommands.h:
```diff
@@ -1071,6 +1071,12 @@ COMMAND("osd pool rmsnap "
+	"name=pool,type=CephPoolname "
+	"name=lower_snapid_bound,type=CephInt,range=0 "
+	"name=upper_snapid_bound,type=CephInt,range=0 ",
+	"remove non-existing snapshot again <snap> from <pool>",
```
Forces re-removal of already removed snapshots in the range ```[lower_snapid_bound, upper_snapid_bound)``` on pool ```<pool>``` in order to cause OSDs to re-trim them after fixing malformed snap_mapper entries.  See bug <bugid> for details.  See also OSD admin socket command fix_malformed_snapmapper_keys.

athanatos commented on src/mon/OSDMonitor.cc:
```diff
@@ -13044,6 +13044,56 @@ bool OSDMonitor::prepare_command_impl(MonOpRequestRef op,
+    }
+
+    pp->set_snap_epoch(pending_inc.epoch);
+    pending_inc.new_pools[pool] = *pp;
```
This isn't necessary, pp is already a pointer to pending_inc.new_pools[pool]

athanatos replied:
This looks right to me otherwise.

Matan-B commented on src/osd/SnapMapper.cc:
```diff
@@ -914,6 +925,15 @@ std::string SnapMapper::convert_legacy_key(
     + "_" + object_suffix;
 }
 
+std::string SnapMapper::convert_malformed_key(
```
This function calls from_raw() which gets const string& - same as previous comment.

Matan-B commented:
Step 2/3 was separated into #53235 since it may stand as a standalone step which unnecessarily relates to the issue addressed in this PR.

Matan-B commented:
Closing, See tracker for resolving this issue.
https://tracker.ceph.com/issues/62596
